### PR TITLE
confile: fix append_unexp_config_line()

### PR DESCRIPTION
**Pull request by brauner**
Reported-by: 2xsec dh48.jeong@samsung.com
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

**Comment by brauner** [comment]
Closes #2695.
//Cc @2xsec :)

**Comment by 2xsec** [comment]
@brauner Thanks for your review. :) I'll check your patch on the our target using the sanitizer then share the result tomorrow. :)

**Comment by brauner** [comment]
Fwiw, I tested it with `-fsanitize=address` and it fixed it for me. :)

**Comment by 2xsec** [comment]
@brauner Before applying your patch, the same issue is reported from sanitizer in your side?

**Comment by brauner** [comment]
Yep:
```
root@wittgenstein|~
> LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libasan.so.5 lxc-start c1
=================================================================
==21438==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x607000000140 at pc 0x7ff2f8ded498 bp 0x7ffdb29dfdc0 sp 0x7ffdb29dfdb0
READ of size 8 at 0x607000000140 thread T0
    #0 0x7ff2f8ded497 in strlcpy ../include/strlcpy.c:46
    #1 0x7ff2f8ded660 in strlcat ../include/strlcat.c:36
    #2 0x7ff2f8c8d08b in append_unexp_config_line /tmp/lxc/src/lxc/confile.c:2057
    #3 0x7ff2f8c8ef73 in parse_line /tmp/lxc/src/lxc/confile.c:2368
    #4 0x7ff2f8d47819 in lxc_file_for_each_line_mmap /tmp/lxc/src/lxc/parse.c:100
    #5 0x7ff2f8c8fe63 in lxc_config_read /tmp/lxc/src/lxc/confile.c:2505
    #6 0x7ff2f8ce5e9e in load_config_locked /tmp/lxc/src/lxc/lxccontainer.c:626
    #7 0x7ff2f8ce602d in do_lxcapi_load_config /tmp/lxc/src/lxc/lxccontainer.c:664
    #8 0x7ff2f8ce60ee in lxcapi_load_config /tmp/lxc/src/lxc/lxccontainer.c:674
    #9 0x7ff2f8d15e0b in lxc_container_new /tmp/lxc/src/lxc/lxccontainer.c:5246
    #10 0x564b7b2b854f in main (/usr/bin/lxc-start+0x254f)
    #11 0x7ff2f89ae09a in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2409a)
    #12 0x564b7b2b89a9 in _start (/usr/bin/lxc-start+0x29a9)

0x607000000142 is located 0 bytes to the right of 66-byte region [0x607000000100,0x607000000142)
allocated by thread T0 here:
    #0 0x7ff2f8eda070 in __interceptor_strdup (/usr/lib/x86_64-linux-gnu/libasan.so.5+0x3b070)
    #1 0x7ff2f8c8eec2 in parse_line /tmp/lxc/src/lxc/confile.c:2363
    #2 0x7ff2f8d47819 in lxc_file_for_each_line_mmap /tmp/lxc/src/lxc/parse.c:100
    #3 0x7ff2f8c8fe63 in lxc_config_read /tmp/lxc/src/lxc/confile.c:2505
    #4 0x7ff2f8ce5e9e in load_config_locked /tmp/lxc/src/lxc/lxccontainer.c:626
    #5 0x7ff2f8ce602d in do_lxcapi_load_config /tmp/lxc/src/lxc/lxccontainer.c:664
    #6 0x7ff2f8ce60ee in lxcapi_load_config /tmp/lxc/src/lxc/lxccontainer.c:674
    #7 0x7ff2f8d15e0b in lxc_container_new /tmp/lxc/src/lxc/lxccontainer.c:5246
    #8 0x564b7b2b854f in main (/usr/bin/lxc-start+0x254f)

SUMMARY: AddressSanitizer: heap-buffer-overflow ../include/strlcpy.c:46 in strlcpy
Shadow bytes around the buggy address:
  0x0c0e7fff7fd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c0e7fff7fe0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c0e7fff7ff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c0e7fff8000: fa fa fa fa 00 00 00 00 00 00 00 00 00 fa fa fa
  0x0c0e7fff8010: fa fa fd fd fd fd fd fd fd fd fd fd fa fa fa fa
=>0x0c0e7fff8020: 00 00 00 00 00 00 00 00[02]fa fa fa fa fa fa fa
  0x0c0e7fff8030: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fff8040: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fff8050: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fff8060: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fff8070: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
```

**Comment by 2xsec** [comment]
@brauner Two kinds of issues are reported from the sanitizer in your side. heap-buffer-overflow and global-buffer-overflow(but, backtrace is the same.) I don't know exactly the two issues are the same. Anyway, I'll check your patch on our side. Thanks. :)

**Comment by brauner** [comment]
Both are fixed by the patch apparently.

**Comment by 2xsec** [comment]
@brauner Good! Thanks. :)

File: src/lxc/confile.c [hunk 1]
```diff
@@ -2050,16 +2052,18 @@ int append_unexp_config_line(const char *line, struct lxc_conf *conf)
...
 		conf->unexpanded_config = tmp;
 		conf->unexpanded_alloced += 1024;
 	}
 
-	(void)strlcat(conf->unexpanded_config, line, conf->unexpanded_alloced);
+	memcpy(conf->unexpanded_config + conf->unexpanded_len, line, linelen);
 	conf->unexpanded_len += linelen;
 	if (line[linelen - 1] != '\n') {
-		(void)strlcat(conf->unexpanded_config, "\n", conf->unexpanded_alloced);
+		memcpy(conf->unexpanded_config + conf->unexpanded_len, "\n", 1);
```
[the remark this answers is not in the document]
Comment (brauner) [reply src/lxc/confile.c]: Sure :)

**Comment by brauner** [comment]
jenkins: test this please

**Pull request merged**